### PR TITLE
WIP: Add haml-lint gem, and fix the problems it finds

### LINKED ISSUE #902: Automatically detect HAML errors
Now that we've eliminated controller tests, it's possible for a simple HAML syntax error to make it through the test suite because the relevant feature test says "oh yes, the string I'm looking for is present in the error page, everything's fine here". This is annoying. We could add controller tests back, but a lighter-weight alternative could be to add a static HAML-checking pass to our test suite, possibly using [haml-lint](https://github.com/brigade/haml-lint).

### PR DESCRIPTION
**Pull request by pozorvlak**
Annoyingly, Rubocop can't auto-correct HAML errors, [and won't gain this ability any time soon](https://github.com/brigade/haml-lint/issues/15).

About 1500 out of the 2633 problems found were "don't use double-quotes where single-quotes will do" warnings. I fixed these using the script and procedure at https://gist.github.com/pozorvlak/af43fed67c698b7e255b2feb91731acb.

The following also ought to be fixable in a (semi)-automated manner:
- [ ] SpaceBeforeScript: The - symbol should have one space separating it from code
- [ ] SpaceBeforeScript: The = symbol should have one space separating it from code
- [ ] RuboCop: Space inside { missing.
- [ ] RuboCop: Space inside } missing.
- [ ] RuboCop: Use `&&` instead of `and`.
- [ ] RuboCop: Use `||` instead of `or`.
- [ ] RuboCop: Surrounding space missing for operator `+`.
- [ ] RuboCop: Operator `?` should be surrounded by a single space.
- [ ] SpaceInsideHashAttributes: Hash attribute should start with one space after the opening brace
- [ ] SpaceInsideHashAttributes: Hash attribute should end with one space before the closing brace
- [ ] ImplicitDiv: `%div.pagination` can be written as `.pagination` since `%div` is implicit

Of the remaining problems, 364 are "LineLength: Line is too long", which I think will have to be fixed manually.

Fixes #902, or will eventually.

### CODE REVIEW

**Comment by cesy** [comment]
It may be worth us adding ReviewNinja or Reviewable for this type of change, to review bit by bit.

**Comment by cesy** [comment]
First two commits are good.

File: app/views/alternate_names/_form.html.haml [hunk 1]
```diff
@@ -13,15 +13,15 @@
...
+    = f.label :name, class: 'control-label col-md-2'
     .col-md-8
-      = f.text_field :name, :class => 'form-control'
+      = f.text_field :name, class: 'form-control'
       
   .form-group
     .form-actions.col-md-offset-2.col-md-8
-      = f.submit 'Save', :class => 'btn btn-primary'
\ No newline at end of file
+      = f.submit 'Save', class: 'btn btn-primary'
```
Comment (cesy) [inline app/views/alternate_names/_form.html.haml]: We may want a separate PR to fix all the missing last lines across the project.

**Comment by cesy** [comment]
Third commit also looks good.

File: app/views/crops/wrangle.html.haml [hunk 2]
```diff
@@ -56,9 +56,9 @@
         - c.scientific_names.each do |s|
           = link_to s.scientific_name, s
           %br/
-      %td= c.requester.present? ? (link_to c.requester, c.requester) : "N/A"
-      - unless @approval_status == "pending"
-        %td= c.creator.present? ? (link_to c.creator, c.creator) : "N/A"
+      %td= c.requester.present? ? (link_to c.requester, c.requester) : 'N/A'
+      - unless @approval_status == 'pending'
+        %td= c.creator.present? ? (link_to c.creator, c.creator) : 'N/A'
```
Comment (cesy) [inline app/views/crops/wrangle.html.haml]: The forward slash won't do anything unexpected, right? I think we're safe here.

File: app/views/layouts/_crowdfunding.html.haml [hunk 3]
```diff
@@ -2,6 +2,6 @@
 - if daysleft > 0
   .crowdfunding-banner
     Help us share free growing information with the world.
-    = link_to "Support our crowdfunding campaign.", "https://www.indiegogo.com/projects/growstuff/x/6079859"
+    = link_to 'Support our crowdfunding campaign.', 'https://www.indiegogo.com/projects/growstuff/x/6079859'
     There are only #{daysleft.to_i} days left.
```
Comment (cesy) [inline app/views/layouts/_crowdfunding.html.haml]: Do we even need this any more, or should we strip it out? Are we likely to do a future similar fundraising effort and reuse it?

File: app/views/plantings/show.html.haml [hunk 4]
```diff
@@ -62,7 +62,7 @@
   %h2 Notes
 
   :growstuff_markdown
-    #{ @planting.description != "" ? @planting.description : "No description given." }
+    #{ @planting.description != '' ? @planting.description : 'No description given.' }
```
Comment (cesy) [inline app/views/plantings/show.html.haml]: Are there any edge cases this might miss?

**Comment by cesy** [comment]
Rest looks fine apart from possibly the bits I've commented on, and the conflicts.

File: app/views/crops/wrangle.html.haml (hunk 2 above)
Comment (pozorvlak) [inline app/views/crops/wrangle.html.haml]:
```
$ irb
2.2.4 :001 > 'N/A'
 => "N/A"
```

Looks fine to me!

**Comment by pozorvlak** [comment]
@cesy Thanks! Would it be easier to merge and then fix any newly-introduced HAML errors in a new commit, or rebase and start again? The latter wouldn't be a problem for me since it's mostly automated.

**Comment by cesy** [comment]
Probably easier to review from the former in this case, though I generally prefer the clean history of the latter.

**Comment by samjewell** [comment]
I'm a visitor to your project, but in passing thought I'd link you to this: http://stackoverflow.com/questions/1836467/is-there-a-performance-gain-in-using-single-quotes-vs-double-quotes-in-ruby

We use double quotes now throughout our haml and our ruby, so that we don't have to change it if we ever need to interpolate. Maybe that would work for your team too...?

**Comment by pozorvlak** [comment]
@samjewell: so you're suggesting disabling the "use of double quotes without interpolation" warning? Yeah, that makes sense to me - @maco, @cesy, @CloCkWeRX, does that sound OK to you?

**Comment by maco** [comment]
fine by me

**Comment by cesy** [comment]
Delaying as @pozorvlak is on holiday.

**Comment by maco** [comment]
So given all that @Br3nda 's been doing, this PR is ...uh...?

**Comment by pozorvlak** [comment]
AFAICT, this PR isn't superseded by Br3nda's work, but I want to start from
scratch and do it again along the lines of her Great Indentation PR. I'll
close this.

**Pull request closed without merging**